lib/TPP/GPU/LinalgToXeGPU.cpp

```diff
@@ -0,0 +1,1402 @@
+      .Case([&](linalg::SubOp subOp) -> std::optional<Value> {
+        assert(operands.size() == 2 && "Invalid number of operands for sub");
+        if (isa<FloatType>(eltType)) {
+          return rewriter
+              .create<arith::SubFOp>(loc, resType, operands[0], operands[1])
+              .getResult();
+        }
+        if (isa<IntegerType>(eltType)) {
+          return rewriter
+              .create<arith::SubIOp>(loc, resType, operands[0], operands[1])
+              .getResult();
+        }
+        // Unhandled type. Bail out.
+        return std::nullopt;
+      })
+      .Case([&](linalg::GenericOp genericOp) -> std::optional<Value> {
+        return lowerGenericOp(genericOp, operands, resType, rewriter);
+      })
+      .Default(
+          [&](Operation *op) -> std::optional<Value> { return std::nullopt; });
+}
+
+// Get static GPU block sizes represented by a surrounding operation
+// like a kernel launch or parallel loop.
+// Returns known block sizes if they are all static or failure, otherwise.
+static FailureOr<SmallVector<int64_t>> getStaticBlockSizes(Operation *op) {
+  if (!op)
+    return failure();
+
+  auto getConstVal = [&](Value val) -> std::optional<int64_t> {
+    if (auto constOp = val.getDefiningOp<arith::ConstantIndexOp>()) {
+      return constOp.value();
+    }
+    return std::nullopt;
+  };
+
+  if (auto launchOp = dyn_cast<gpu::LaunchOp>(op)) {
+    auto sizeX = getConstVal(launchOp.getBlockSizeX());
+    auto sizeY = getConstVal(launchOp.getBlockSizeY());
+    auto sizeZ = getConstVal(launchOp.getBlockSizeZ());
+    if (!sizeX || !sizeY || !sizeZ)
+      return failure();
+
+    return SmallVector<int64_t>{*sizeX, *sizeY, *sizeZ};
+  }
+
+  // TODO: Remove when the lowering only occurs within a gpu.launch op.
+  //       Manually computing this is brittle and duplicated parallel
+  //       loops to gpu conversion.
+  if (auto blockLoop = dyn_cast<scf::ParallelOp>(op)) {
+    auto gridLoop = blockLoop->getParentOfType<scf::ParallelOp>();
+
+    // Blocks or number of threads are represented by the first parallel loop
+    // nested within another parallel loop.
+    //
+    // Fail if there is no outer parallel loop or current loop is nested more
+    // than once.
+    if (!gridLoop || (gridLoop->getParentOfType<scf::ParallelOp>())) {
+      return failure();
+    }
+
+    SmallVector<int64_t> blockSizes;
+    for (auto [lb, ub, step] :
+         llvm::zip_equal(blockLoop.getLowerBound(), blockLoop.getUpperBound(),
+                         blockLoop.getStep())) {
+      auto lbVal = getConstVal(lb);
+      auto ubVal = getConstVal(ub);
+      auto stepVal = getConstVal(step);
+      if (!lbVal || !ubVal || !stepVal)
+        return failure();
+
+      int64_t blockSize = (*ubVal - *lbVal) / *stepVal;
+
+      // Assume that at least one subgroup is created for the given
+      // dimension. Otherwise, outlining will fail anyway.
+      blockSizes.push_back(blockSize < 0 ? 1 : blockSize);
```

Comment:
When would this ever be `< 0`?

Reply:
In case of somewhat ill defined IR like `lb > ub` (overall valid but wrong in this context) or one of them negative.
I think returning failure might be a bit better here.

Reply:
or even an assert. This sounds like broken IR to me.

Reply:
Tweaked the check.

Reply:
Went with failure like in the `if (blockSizes.size() > 3)` case. Further GPU passes will fail anyway.